nucypher/blockchain/eth/actors.py

```diff
@@ -1643,84 +1653,201 @@ def __init__(self,
                  client_password: str = None,
                  *args, **kwargs):
         super().__init__(checksum_address=checksum_address, *args, **kwargs)
-        self.worklock_agent = ContractAgency.get_agent(WorkLockAgent, registry=self.registry)
+        self.worklock_agent = ContractAgency.get_agent(WorkLockAgent, registry=self.registry)  # type: WorkLockAgent
         self.staking_agent = ContractAgency.get_agent(StakingEscrowAgent, registry=self.registry)
         self.economics = EconomicsFactory.get_economics(registry=self.registry)
 
         if is_transacting:
             self.transacting_power = TransactingPower(password=client_password, account=checksum_address)
             self.transacting_power.activate()
 
-    def _ensure_bidding_is_open(self) -> None:
+        self._all_bonus_bidders = None
+
+    def _ensure_bidding_is_open(self, message: str = None) -> None:
         highest_block = self.worklock_agent.blockchain.w3.eth.getBlock('latest')
         now = highest_block['timestamp']
-        start = self.worklock_agent.start_date
-        end = self.worklock_agent.end_date
+        start = self.worklock_agent.start_bidding_date
+        end = self.worklock_agent.end_bidding_date
         if now < start:
-            raise self.BiddingIsClosed(f'Bidding does not open until {maya.MayaDT(start).slang_date()}')
-        if now > end:
-            raise self.BiddingIsClosed(f'Bidding closed at {maya.MayaDT(end).slang_date()}')
+            message = message or f'Bidding does not open until {maya.MayaDT(start).slang_date()}'
+            raise self.BiddingIsClosed(message)
+        if now >= end:
+            message = message or f'Bidding closed at {maya.MayaDT(end).slang_date()}'
+            raise self.BiddingIsClosed(message)
 
     def _ensure_bidding_is_closed(self, message: str = None) -> None:
         highest_block = self.worklock_agent.blockchain.w3.eth.getBlock('latest')
         now = highest_block['timestamp']
-        end = self.worklock_agent.end_date
+        end = self.worklock_agent.end_bidding_date
         if now < end:
-            message = message or f"Bidding does not close until {end}"
+            message = message or f"Bidding does not close until {maya.MayaDT(end).slang_date()}"
             raise self.BiddingIsOpen(message)
 
+    def _ensure_cancellation_window(self, ensure_closed: bool = True, message: str = None) -> None:
+        highest_block = self.worklock_agent.blockchain.w3.eth.getBlock('latest')
+        now = highest_block['timestamp']
```

Comment:
Consider encapsulating `get_blocktime` in a method this might also help with future upstream API changes and test mocks.

Reply:
✔️